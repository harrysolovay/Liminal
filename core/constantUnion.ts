import { partition } from "@std/collections"
import { Ty } from "./Ty.ts"

export type constantUnion<M extends Array<number | string>> = ReturnType<typeof constantUnion<M>>

export function constantUnion<M extends Array<number | string>>(
  ...members: M
): Ty<M[number], never> {
  const [strings, numbers] = partition(members, (v) => typeof v === "string")
  return Ty((description) => {
    return {
      anyOf: [
        ...strings.length
          ? [{
            type: "string",
            enum: strings,
          }]
          : [],
        ...numbers.length
          ? [{
            type: "number",
            enum: numbers,
          }]
          : [],
      ],
      description,
    }
  })
}
