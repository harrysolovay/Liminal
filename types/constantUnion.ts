import { partition } from "@std/collections"
import { Ty } from "./Ty.ts"

// TODO: use single `enum` type in cases where there aren't both numbers and strings
export function constantUnion<M extends Array<number | string>>(
  ...members: M
): Ty<M[number], never, false> {
  const [strings, numbers] = partition(members, (v) => typeof v === "string")
  return Ty(() => ({
    anyOf: [
      ...strings.length
        ? [{
          type: "string",
          enum: strings,
        }]
        : [],
      ...numbers.length
        ? [{
          type: "number",
          enum: numbers,
        }]
        : [],
    ],
  }), false)
}
