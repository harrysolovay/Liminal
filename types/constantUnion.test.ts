import { constantUnion } from "./constantUnion.ts"
import { assertTySnapshot } from "test_util"

Deno.test("constant union", async (t) => {
  await assertTySnapshot(t, constantUnion("Hello", "world!", 42, 420))
})
