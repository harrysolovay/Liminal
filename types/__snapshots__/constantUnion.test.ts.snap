export const snapshot = {};

snapshot[`constant union 1`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      anyOf: [
        {
          enum: [
            "Hello",
            "world!",
          ],
          type: "string",
        },
        {
          enum: [
            42,
            420,
          ],
          type: "number",
        },
      ],
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`constant union 2`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      anyOf: [
        {
          enum: [
            "Hello",
            "world!",
          ],
          type: "string",
        },
        {
          enum: [
            42,
            420,
          ],
          type: "number",
        },
      ],
      description: "Description one.",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`constant union 3`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      anyOf: [
        {
          enum: [
            "Hello",
            "world!",
          ],
          type: "string",
        },
        {
          enum: [
            42,
            420,
          ],
          type: "number",
        },
      ],
      description: "Description two. Description one.",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;
